File: reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    # iteration : O(n), O(n)
    def reverseList(self, head: ListNode) -> ListNode:
        tail = None
        while head:
            tail, tail.next, head = head, tail, head.next
        return tail

    # recursion : O(n), O(n)
    def reverseList(self, head: ListNode) -> ListNode:
        def reverse(head: ListNode, tail: ListNode = None):
            if not head:
                return tail
            post, head.next = head.next, tail
            return reverse(post, head)
        return reverse(head)

File: reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode node = null;
        ListNode next;
        while (head != null) {
            next = head.next;
            node = new ListNode(head.val, node);
            head = next;
        }

        return node;
    }
}